src/util.py

```diff
@@ -0,0 +1,320 @@
+# TODO: Move this into dmutils
```

Comment:
I'm assuming this was just copied from another place, so I'm not going to review it in detail. We can do that after it has migrated to `dml-utils` and then we can do a review of the file as a whole. I don't want to get changes lost here because we move some other copy to `dml-utils`.

Reply:
mostly copied and just refactored a bit, yeah. that sounds good. its also gonna change quite a bit with some incoming changes from natalie and mohammad.